### ISSUE TITLE
Prefer '*.hpp' over '*.h' (opionated)

### ISSUE THREAD

**Issue by daixtrose** [copperspice/cs_libguarded#16]
## TL;DR: 

Please revert https://github.com/copperspice/cs_libguarded/commit/98b9c17e76b9dc639f2936146e2201f52c5b2f44

## Discussion

IIRC there were really good reasons to have C++ header files end with `.hpp` rather than `.h`. Some had to do with IDE discriminating between C and C++. I believe the [Boost Community](https://boost.org) went through this discussion some decades ago, but I can't find the original discussion quickly. Please find other opinions below

### Pro 

- https://stackoverflow.com/questions/152555/h-or-hpp-for-your-class-definitions
- https://www.boost.org/users/faq.html
  > Why do Boost headers have a .hpp suffix rather than .h or none at all? File extensions communicate the "type" of the file, both to humans and to computer programs. The '.h' extension is used for C header files, and therefore communicates the wrong thing about C++ header files. Using no extension communicates nothing and forces inspection of file contents to determine type. Using '.hpp' unambiguously identifies it as C++ header file, and works well in actual practice. (Rainer Deyke)
- https://llvm.org/docs/CodingStandards.html#file-headers

### Contra
 
- [Bjarne is against my proposal](https://github.com/isocpp/CppCoreGuidelines/issues/686#issuecomment-242524029)  :frowning_face: 
- [Google Coding Guidelines](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)

**Comment by bgeller** [comment]
As explained in a comment on the commit, this was done to conform with the rest of the CopperSpice code base. We understand developers have different preferences, however consistency is extremely important. Looking at your pro and con list sort of shows that is no formal agreement.
